src/AspNetCore/Src/applicationmanager.cxx

```diff
@@ -197,72 +222,194 @@ APPLICATION_MANAGER::RecycleApplication(
     HRESULT          hr = S_OK;
     APPLICATION_INFO_KEY  key;
     DWORD            dwPreviousCounter = 0;
+    APPLICATION_INFO_HASH* table = NULL;
+    CONFIG_CHANGE_CONTEXT context;
 
     hr = key.Initialize(pszApplicationId);
     if (FAILED(hr))
     {
         goto Finished;
     }
+
+    table = new APPLICATION_INFO_HASH();
+
+    if(table == NULL)
+    {
+        hr = E_OUTOFMEMORY;
+        goto Finished;
+    }
+
+    //
+    // few application expected, small bucket size for hash table
+    //
+    if (FAILED(hr = table->Initialize(17 /*prime*/)))
+    {
+        goto Finished;
+    }
+
+    context.pstrPath = pszApplicationId;
+
     AcquireSRWLockExclusive(&m_srwLock);
     dwPreviousCounter = m_pApplicationInfoHash->Count();
 
-    m_pApplicationInfoHash->DeleteKey(&key);
+    // We don't want to hold the lock for long time as it will block all incoming requests
+    // Make a shallow copy of existing hashtable as we may nee to remove nodes
+    // This also make sure application shutdown will not be called inside the lock
+    m_pApplicationInfoHash->Apply(APPLICATION_INFO_HASH::ReferenceCopyToTable, static_cast<PVOID>(table));
+    DBG_ASSERT(dwPreviousCounter == table->Count());
+    
+    // Removed the applications which are impacted by the configurtion change
+    m_pApplicationInfoHash->DeleteIf(FindConfigChangedApplication, (PVOID)&context);
 
-    if(dwPreviousCounter != m_pApplicationInfoHash->Count())
+    if (dwPreviousCounter != m_pApplicationInfoHash->Count())
     {
-        // Application got recycled. Log an event
-        STACK_STRU(strEventMsg, 256);
-        if (SUCCEEDED(strEventMsg.SafeSnwprintf(
-            ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG,
-            pszApplicationId)))
+        if (m_hostingModel == HOSTING_IN_PROCESS)
```

Comment:
For in-proc scenario, it looks like we don't actually need to perform the shallow copy of the hashtable? Since we don't need to modify m_pApplicationInfoHash at the end, all we need to know is whether the configure change affects one of the applications in the hash table, and then we only need to recycle w3wp upon that? Maybe we could have a small optimization here:?